[ArgPromotion] Perform alias analysis on actual arguments of Calls

hazzlim opened the pull request:
Teach Argument Promotion to perform alias analysis on actual arguments of Calls to a Function, to try to prove that all Calls to the Function do not modify the memory pointed to by an argument. This surfaces more opportunities to perform Argument Promotion in cases where simply looking at a Function's instructions is insufficient to prove that the pointer argument is not invalidated before all loads from it.

hazzlim committed:
[ArgPromotion] Perform alias analysis on actual arguments of Calls

Teach Argument Promotion to perform alias analysis on actual arguments
of Calls to a Function, to try to prove that all Calls to the Function
do not modify the memory pointed to by an argument. This surfaces more
opportunities to perform Argument Promotion in cases where simply
looking at a Function's instructions is insufficient to prove that the
pointer argument is not invalidated before all loads from it.

efriedma-quic commented on llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:
```diff
@@ -745,6 +748,33 @@ static bool areTypesABICompatible(ArrayRef<Type *> Types, const Function &F,
+  for (User *U : F->users()) {
+    auto *Call = dyn_cast<CallInst>(U);
+    if (!Call)
+      continue;
```
If we find an unexpected use of the function, should we bail?

efriedma-quic commented on llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:
```diff
@@ -745,6 +748,33 @@ static bool areTypesABICompatible(ArrayRef<Type *> Types, const Function &F,
+    Value *ArgOp = Call->getArgOperand(ArgNo);
+    assert(ArgOp->getType()->isPointerTy() && "Argument must be Pointer Type!");
+
+    MemoryLocation Loc = MemoryLocation::getForArgument(Call, ArgNo, nullptr);
```
Can we construct a more precise memory location based on the actual loads in the callee?  This is basically equivalent to `MemoryLocation::getBeforeOrAfter(ArgOp)`

hazzlim replied:
If I understand correctly, I don't think that we can do this - because BasicAA does not support inter-procedural analysis, we cannot pass a Loc derived from a load in the callee to the AAR query based on the caller (where the Call to F is).
https://github.com/llvm/llvm-project/blob/6c143a86cddbc6d0431dd643bfc7d4f017042512/llvm/lib/Analysis/BasicAliasAnalysis.cpp#L925-L929

efriedma-quic replied:
I mean, use the pointer from the caller, but compute the size of the memory region using the loads in the callee.

hazzlim replied:
Ah I see, apologies I was misunderstanding you before. I have added this in [Address review 3](https://github.com/llvm/llvm-project/pull/106216/commits/6b9b01890f86389212cbe762e43bf896215dd9dd)

If I understand correctly we still run into an issue because we are doing an alias query on a CallInst, and because BasicAA is inherently non-interprocedural it disregards the Size of the Loc and uses getBeforeOrAfter anyway when checking if the pointer aliases any of the Call's arguments:
https://github.com/llvm/llvm-project/blob/6c143a86cddbc6d0431dd643bfc7d4f017042512/llvm/lib/Analysis/BasicAliasAnalysis.cpp#L973-L978

efriedma-quic commented on llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:
```diff
@@ -745,6 +748,33 @@ static bool areTypesABICompatible(ArrayRef<Type *> Types, const Function &F,
+// Try to prove that all Calls to F do not modify the memory pointed to by Arg.
+// This can provide us with more opportunities to perform Argument Promotion in
+// cases where simply looking at a Function's instructions is insufficient to
+// prove that the pointer argument is not invalidated before all loads from it.
```
So if I'm following correctly, you're doing basically the same analysis in two different ways.  The first way is using local to the caller: if the pointed-to memory isn't modified at all, we're fine.  
can prove a lack of aliasing, we're done.  The second way is local to the callee: if the argument memory can't ever be modified before it's loaded, no matter what the arguments are, we're fine.

In theory, you could do an interprocedural analysis... but it would be expensive to compute, and ArgumentPromotion doesn't have infrastructure for that.  So I guess this is fine, but it could use better comments explaining that you're using two different approaches to compute the same thing.

Also, is there some reason callDoesNotModifyArg() needs to run before we reach the relevant bit of findArgParts()?  If the two checks were together, it would be more obvious they're related.

hazzlim replied:
Yes exactly - ideally we would be able to do a more complete inter-procedural analysis, but in lieu of that this seemed like a reasonable way to get ArgPromotion to handle the motivating case (which comes from the SPEC bwaves benchmark).

I've moved the call to the relevant place in `findArgParts`, and tried to make things clearer in the comments.

Thank you for reviewing the PR :)

hazzlim committed:
Address review comments

- Bail on unexpected (non CallInst) use of function
- Move checks into findArgParts to improve clarity and only run when
  necessary
- Remove FIXMEs from test

MDevereau commented on llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:
```diff
@@ -485,11 +485,39 @@ static bool allCallersPassValidPointerForArgument(
 
+// Try to prove that all Calls to F do not modify the memory pointed to by Arg,
+// using alias analysis local to each caller of F.
+static bool isArgUnmodifiedByAllCalls(Function *F, unsigned ArgNo,
+                                      FunctionAnalysisManager &FAM) {
```
Can we pass just in an `Argument*` here and use `Argument->getParent()` and `Argument->getArgNo()` to simplify the function signature?

hazzlim replied:
Good point, that's much cleaner - I've made this change in commit [Address review comments 2](https://github.com/llvm/llvm-project/pull/106216/commits/aa809e9dc8a9bf5b5fc0cdc6f56549bd7806879d)

MDevereau commented on llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:
```diff
@@ -485,11 +485,39 @@ static bool allCallersPassValidPointerForArgument(
+      return false;
+  }
+
+  // All Users are Calls which do not modify the Arg.
```
This comment, the one on line 492, the one on line 488/499, and the function name all pretty much say the same thing. Maybe delete the ones in the function body

hazzlim replied:
Removed redundant comments in [Address review comments 2](https://github.com/llvm/llvm-project/pull/106216/commits/aa809e9dc8a9bf5b5fc0cdc6f56549bd7806879d)

MDevereau commented on llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:
```diff
@@ -485,11 +485,39 @@ static bool allCallersPassValidPointerForArgument(
+    if (!Call)
+      return false;
+
+    Value *ArgOp = Call->getArgOperand(ArgNo);
+    assert(ArgOp->getType()->isPointerTy() && "Argument must be Pointer Type!");
```
I don't think its possible for Arg to not be a pointer type at this stage, since `PointerArgs` in `promoteArguments`. I think this check can be deleted.

hazzlim replied:
Fair point - I've removed this assert in [Address review comments 2](https://github.com/llvm/llvm-project/pull/106216/commits/aa809e9dc8a9bf5b5fc0cdc6f56549bd7806879d)

hazzlim committed:
Address review comments 2

- Remove redundant comments
- Refactor function signature
- Remove unnecessary assert

hazzlim committed:
Address review 3

- Compute the size of the memory region using the loads in the callee

hazzlim commented:
Polite ping on this :)

hazzlim committed:
Address review 4

- Fix incorrect calculation of the size of the memory region accessed by
  the loads through Arg.

efriedma-quic commented on llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:
```diff
@@ -716,10 +744,22 @@ static bool findArgParts(Argument *Arg, const DataLayout &DL, AAResults &AAR,
+  LocationSize Size = LocationSize::precise(0);
+  for (LoadInst *Load : Loads) {
+    Size = Size.unionWith(MemoryLocation::get(Load).Size);
+  }
```
I think the case where it would matter is if you pass in the same pointer twice (or a related pointer).  Which, umm.... I guess we can't prove aliasing in that case anyway.  I mean, it could matter in the future if getModRefInfo becomes more powerful, but I don't think getModRefInfo can prove something like that at the moment.

Sorry, I didn't realize when I made the suggestion.

efriedma-quic commented on llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:
```diff
@@ -716,10 +745,16 @@ static bool findArgParts(Argument *Arg, const DataLayout &DL, AAResults &AAR,
+  // If we can determine that no call to the Function modifies the memory region
+  // accessed through Arg, through alias analysis using actual arguments in the
+  // callers, we know that it is guaranteed to be safe to promote the argument.
+  if (isArgUnmodifiedByAllCalls(Arg, LocationSize::precise(Offset), FAM))
```
I'm not sure this is quite right; can an Offset be negative?

hazzlim replied:
Ah yes - I was mistakenly thinking that we bail when any load has a negative offset from Arg, but this is only the case for loads in the entry block that are guaranteed to execute. And I think what I've done here is in fact incorrect for all cases where the lowest offset is non-zero.

Given that it is not really possible to test this at this stage, I have reverted to not using any Size information if you think that seems appropriate?

efriedma-quic reviewed (commented):
I'm a little concerned that because we can't write tests, it'll be harder to track down issues later if there's something wrong with the size computation.

hazzlim committed:
Remove Size calculation

efriedma-quic reviewed (approved):
LGTM

nikic commented on llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:
```diff
@@ -485,11 +485,36 @@ static bool allCallersPassValidPointerForArgument(
+  for (User *U : Arg->getParent()->users()) {
+
+    // Bail if we find an unexpected (non CallInst) use of the function.
+    auto *Call = dyn_cast<CallInst>(U);
```
Why is this checking for `CallInst` rather than `CallBase`? This should also work for invokes, right?

And I think then you could also replace it with an assert, as the function only being used as CallBase callee is a precondition of trying the transform in the first place.

hazzlim replied:
Apologies, I was perhaps a bit trigger happy with committing this. I have put up #110335 to address this.
